Mono.Cecil.Cil/MethodBody.cs

```diff
@@ -163,10 +165,35 @@ protected override void OnSet (VariableDefinition item, int index)
 
 		protected override void OnRemove (VariableDefinition item, int index)
 		{
+			UpdateVariableIndices (index + 1, -1, index);
 			item.index = -1;
+		}
 
-			for (int i = index + 1; i < size; i++)
-				items [i].index = i - 1;
+		void UpdateVariableIndices (int startIndex, int offset, int indexToRemove = -1)
+		{
+			for (int i = startIndex; i < size; i++)
+				items [i].index = i + offset;
+
+			var debug_info = method == null ? null : method.debug_info;
+			if (debug_info == null || debug_info.Scope == null)
+				return;
+
+			foreach (var scope in debug_info.GetScopes ()) {
+				if (scope.HasVariables) {
+					var variables = scope.Variables;
+					for (int i = 0; i < variables.Count; i++) {
+						if (variables [i].Index == indexToRemove) {
+							variables.RemoveAt (i);
+						}
+					}
+
+					foreach (var variable in variables) {
+						if (variable.index.index.HasValue && variable.index.index.Value >= startIndex) {
```

Comment:
Because this looks a bit odd, I think we want to comment as to why we're doing this. In that case my understanding is that you want to update the indexes of variables that have not been resolved into a VariableDefinition, while a VariableIndex can represent both the resolved and not resolved case.

Reply:
Hopefully the new version is more readable (and I also added a comment) - and it now also handles removal of resolved variable debug infos.